# DTSC 580: Data Manipulation
## Assignment: LEGO Project
### Name: 

![LegoSchema](Lego_Schema.PNG)

## Overview

In this assignment, you will load the various tables from the [Rebrickable](https://rebrickable.com/) website and perform analysis on the data.  Rebrickable is a fantastic website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.  There is no data dictionary that I am aware of, but you should be able to understand this data by reviewing the files and schema and visiting their [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation.  **Note that you must use the files downloaded from Brightspace for this assignment and not download new files from the Rebrickable website.**

## Assignment

In this assignment, you will create a number of functions to complete certain tasks.  This is not needed if you were working on this project on your own, but since we want to see if you can select the data via Pandas/Numpy, we will use those functions in the automatic grading in CodeGrade.  Without this, students could try to look up the answers in the `csv` files instead of wrangling the data with Python/Pandas, which would defeat the purpose of this class.

In addition, we would strongly suggest that you merge data sets as appropriate for the given task.  For example, you might merge the `sets` and `themes` DataFrames to answer one question, and then merge `sets`, `inventories`, `inventory_minifigs`, and `minifigs` to answer another.  This will make your code much easier to write and will be key in working with data across multiple tables.  In fact, I was able to create each function with only one line of code because of the merging of certain data sets.  Also note that you may not use all the tables in this assignment.  This is also an important skill for a Data Scientist to know when to use data and what data is unnecessary.

When complete, save your notebook as `lego.ipynb` and submit to CodeGrade for automatic grading.

## Note

<u>Show Work</u>

Remember that you must show your work.  Students submissions are spot checked manually throughout the term to verify that they are not hard coding the answer from looking only in the file or in CodeGrade's expected output.  If this is seen, the student's answer will be manually marked wrong and their grade will be changed to reflect this. 

For example, if the answer to Q1, the mean of a specific column, is 22:
```
# correct way
Q1 = df['column_name'].mean()

# incorrect way
Q1 = 22 
```

Let's start out by importing some standard imports.

In [1]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)
import warnings
warnings.filterwarnings("ignore")

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [3]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [5]:
print("Inventories Dataset")
print(inventories.head())
print(inventories.info())

print("\nInventory Sets Dataset")
print(inventory_sets.head())
print(inventory_sets.info())

print("\nSets Dataset")
print(sets.head())
print(sets.info())

print("\nThemes Dataset")
print(themes.head())
print(themes.info())

print("\nInventory Minifigs Dataset")
print(inventory_minifigs.head())
print(inventory_minifigs.info())

print("\nMinifigs Dataset")
print(minifigs.head())
print(minifigs.info())

print("\nInventory Parts Dataset")
print(inventory_parts.head())
print(inventory_parts.info())

print("\nPart Categories Dataset")
print(part_categories.head())
print(part_categories.info())

print("\nParts Dataset")
print(parts.head())
print(parts.info())

print("\nColors Dataset")
print(colors.head())
print(colors.info())

print("\nPart Relationships Dataset")
print(part_relationships.head())
print(part_relationships.info())

print("\nElements Dataset")
print(elements.head())
print(elements.info())

Inventories Dataset
   id  version set_num
0   1        1  7922-1
1   3        1  3931-1
2   4        1  6942-1
3  15        1  5158-1
4  16        1   903-1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31587 entries, 0 to 31586
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   id       31587 non-null  int64 
 1   version  31587 non-null  int64 
 2   set_num  31587 non-null  object
dtypes: int64(2), object(1)
memory usage: 740.4+ KB
None

Inventory Sets Dataset
   inventory_id  set_num  quantity
0            35  75911-1         1
1            35  75912-1         1
2            39  75048-1         1
3            39  75053-1         1
4            50   4515-1         1
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3413 entries, 0 to 3412
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   inventory_id  3413 non-null   int64 
 1   set_num       3413 no

## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [7]:
def data_count(df):
    return df.shape[0]

sets_total_rows = data_count(sets)
print(sets_total_rows)

themes_total_rows = data_count(themes)
print(themes_total_rows)

18576
444


**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [9]:
def data_count(df):
    return df.shape[0]

sets_total_rows = data_count(sets)

print(sets_total_rows)

18576


**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [11]:
Q2 = round(sets['num_parts'].mean())

print(Q2)

161


**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [13]:
Q3 = int(sets['num_parts'].median())

print(Q3)

37


You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [15]:
Q4 = sets['num_parts'].max()

print(Q4)

11695


**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [17]:
Q5 = sets[sets['num_parts'] == Q4]

print(Q5)

      set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695


**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [19]:
def select_set_row(set_num):
    return sets[sets['set_num'] == set_num]

**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [21]:
def select_set_numparts(num_parts):

    filtered_df = sets[sets['num_parts'] == num_parts]

    sorted_df = filtered_df.sort_values(by=['year', 'set_num'])
    return sorted_df

**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [23]:
Q7_check = select_set_numparts(0)

print(Q7_check.shape[0])

3146


**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [25]:
Q8A = sets['year'].min()

print(Q8A)

1949


In [27]:
Q8B = sets['year'].max()

print(Q8B)

2022


**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.

In [29]:
def select_set_year(year):

    filtered_df = sets[sets['year'] == year]

    sorted_df = filtered_df.sort_values(by='set_num')
    return sorted_df

**Q10:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [61]:
def theme_by_year(year):

    sets_in_year = sets[sets['year'] == year]
    
    merged_df = pd.merge(sets_in_year, themes, left_on='theme_id', right_on='id')
    
    themes_in_year = merged_df[['theme_id', 'name_y']].drop_duplicates(subset='theme_id')
    
    themes_in_year.columns = ['id', 'name_themes']
    
    themes_in_year = themes_in_year.sort_values(by='id')
    
    themes_in_year = themes_in_year.reset_index(drop=True)
    
    return themes_in_year

Q10_check = theme_by_year(1960)

print(Q10_check)

    id   name_themes
0  371  Supplemental
1  497         Books
2  513       Classic


**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

**Q11:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [57]:
def theme_by_name(theme_name):

    theme_ids = themes[themes['name'] == theme_name]['id']
    
    sets_in_theme = sets[sets['theme_id'].isin(theme_ids)]
    
    merged_df = pd.merge(sets_in_theme, themes, left_on='theme_id', right_on='id')
    
    result_df = merged_df[['set_num', 'name_x', 'year', 'num_parts', 'theme_id', 'name_y']]
    result_df.columns = ['set_num', 'name_sets', 'year', 'num_parts', 'theme_id', 'name_themes']
    
    result_df = result_df.sort_values(by='set_num').reset_index(drop=True)
    
    return result_df

Q11_check = theme_by_name("Gear")

print(Q11_check.shape[0])

print(Q11_check.head())

1904
          set_num                              name_sets  year  num_parts  \
0         11138-1                Bionicle Barraki Wallet  2008          0   
1  119645111436-1    Classic Space Astronaut Plush (Red)  2021          0   
2  119645111504-1   Classic Space Astronaut Plush (Blue)  2021          0   
3  119645111740-1  Classic Space Astronaut Plush (White)  2021          0   
4  119645111818-1   Classic Space Astronaut Plush (Pink)  2021          0   

   theme_id name_themes  
0       501        Gear  
1       501        Gear  
2       501        Gear  
3       501        Gear  
4       501        Gear  


**Code Check:** Call the `theme_by_name()` function using the theme name `Gear`.  You should return 1,904 sets that match that theme name.

In [59]:
Q11_check = theme_by_name("Gear")

print(Q11_check.shape[0])

1904


**Q12:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [63]:
def theme_by_setnum(set_num):

    theme_id = sets[sets['set_num'] == set_num]['theme_id'].values[0]
    
    theme_name = themes[themes['id'] == theme_id]['name'].values[0]
    
    return theme_name

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [72]:
def theme_by_setnum(set_num):

    if set_num in sets['set_num'].values:

        theme_id = sets[sets['set_num'] == set_num]['theme_id'].values[0]
        
        theme_name = themes[themes['id'] == theme_id]['name'].values[0]
        
        return theme_name
    else:
        return "Set number not found in the dataset"

set_num_to_check = "75257"
theme_name = theme_by_setnum(set_num_to_check)

# Check for set numbers that start with "752" (or any prefix you want to explore)
sets[sets['set_num'].str.startswith("752")]

print(f"The theme name for set number {set_num_to_check} is: {theme_name}")

The theme name for set number 75257 is: Set number not found in the dataset


**Millennium Falcon Prep Exercise:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You will use this information for later questions.

In [74]:
millennium_falcon_sets = sets[sets['name'] == 'Millennium Falcon']

largest_millennium_falcon_set = millennium_falcon_sets[millennium_falcon_sets['num_parts'] == millennium_falcon_sets['num_parts'].max()]

print(largest_millennium_falcon_set)

       set_num               name  year  theme_id  num_parts
12884  75192-1  Millennium Falcon  2017       171       7541


**Q13:** 
- How many minifigs are contained in the set from the `Millennium Falcon Prep Exercise`?  Save this as `Q13` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [76]:
largest_millennium_falcon_set_num = largest_millennium_falcon_set['set_num'].values[0]

inventory_id = inventories[inventories['set_num'] == largest_millennium_falcon_set_num]['id'].values[0]

Q13 = inventory_minifigs[inventory_minifigs['inventory_id'] == inventory_id]['quantity'].sum()

print(Q13)

8


**Q14:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [78]:
def get_minifigs(set_num, version_num):

    inventory_id = inventories[(inventories['set_num'] == set_num) & (inventories['version'] == version_num)]['id'].values[0]
    
    minifigs_in_set = inventory_minifigs[inventory_minifigs['inventory_id'] == inventory_id]
    
    minifigs_details = pd.merge(minifigs_in_set, minifigs, left_on='fig_num', right_on='fig_num')
    
    result_df = minifigs_details[['set_num', 'fig_num', 'name', 'num_parts']]
    result_df.columns = ['set_num', 'fig_num', 'fig_name', 'fig_num_parts']
    
    result_df = result_df.sort_values(by='fig_num').reset_index(drop=True)
    
    return result_df

**Q15:** Using the above function, output information about the minifigs from the set discussed in the `Millennium Falcon Prep Exercise`.  Save this output as `Q15`.

In [82]:
def get_minifigs(set_num, version_num):

    inventory_id = inventories[(inventories['set_num'] == set_num) & (inventories['version'] == version_num)]['id'].values[0]

    minifigs_in_set = inventory_minifigs[inventory_minifigs['inventory_id'] == inventory_id]
    
    minifigs_in_set = pd.merge(minifigs_in_set, inventories[['id', 'set_num']], left_on='inventory_id', right_on='id')
    
    minifigs_details = pd.merge(minifigs_in_set, minifigs, left_on='fig_num', right_on='fig_num')

    result_df = minifigs_details[['set_num', 'fig_num', 'name', 'num_parts']]
    result_df.columns = ['set_num', 'fig_num', 'fig_name', 'fig_num_parts']
    
    result_df = result_df.sort_values(by='fig_num').reset_index(drop=True)
    
    return result_df

largest_millennium_falcon_set_num = largest_millennium_falcon_set['set_num'].values[0]

Q15 = get_minifigs(largest_millennium_falcon_set_num, 1)

print(Q15)

   set_num     fig_num                                           fig_name  \
0  75192-1  fig-001714                              Chewbacca, Dark Brown   
1  75192-1  fig-001810                             Finn, Black Undershirt   
2  75192-1  fig-002057                                 Rey, Dark Tan Robe   
3  75192-1  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
4  75192-1  fig-002544                                               BB-8   
5  75192-1  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
6  75192-1  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
7  75192-1  fig-002551                               Han Solo, Old, Angry   

   fig_num_parts  
0              3  
1              4  
2              4  
3              3  
4              2  
5              4  
6              4  
7              4  


**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [84]:
Q15_version1 = get_minifigs("10217-1", 1)
print("Version 1 Minifigs:")
print(Q15_version1)
print(f"Total Minifigs in Version 1: {Q15_version1.shape[0]}")

Q15_version2 = get_minifigs("10217-1", 2)
print("\nVersion 2 Minifigs:")
print(Q15_version2)
print(f"Total Minifigs in Version 2: {Q15_version2.shape[0]}")

Version 1 Minifigs:
    set_num     fig_num                                           fig_name  \
0   10217-1  fig-002229                   Skeleton (Standard) (3626b Head)   
1   10217-1  fig-006075                 Rubeus Hagrid, Dark Brown Overcoat   
2   10217-1  fig-006084  Lucius Malfoy, Black Robes and Cape, Death Eat...   
3   10217-1  fig-006130  Harry Potter, Open Dark Blue Jacket over White...   
4   10217-1  fig-006138                                    Fenrir Greyback   
5   10217-1  fig-006162  Ron Weasley, Red Tartan Sweater Vest, Reddish ...   
6   10217-1  fig-006163          Fred / George Weasley in Dark Orange Suit   
7   10217-1  fig-006164                                 Garrick Ollivander   
8   10217-1  fig-006165    Goblin, Black Jacket and Light Bluish Gray Legs   
9   10217-1  fig-006166        Goblin, Black Jacket and Reddish Brown Legs   
10  10217-1  fig-006167  Hermione Granger, Light Bluish Gray Sweater wi...   

    fig_num_parts  
0               6  
1  

**Q16:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [86]:
def minifigs_from_themes(theme_id):

    set_nums = sets[sets['theme_id'] == theme_id]['set_num']
    
    inventory_ids = inventories[inventories['set_num'].isin(set_nums)]['id']
    
    minifigs_in_theme = inventory_minifigs[inventory_minifigs['inventory_id'].isin(inventory_ids)]
    
    minifigs_details = pd.merge(minifigs_in_theme, minifigs, left_on='fig_num', right_on='fig_num')
    
    result_df = minifigs_details[['fig_num', 'name']].drop_duplicates().sort_values(by='fig_num')
    
    result_df.columns = ['fig_num', 'fig_name']
    
    result_df = result_df.reset_index(drop=True)
    
    return result_df

**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in the `Millennium Falcon Prep Exercise`.  There should be 149 minifigs output for that respective theme.

In [88]:
millennium_falcon_theme_id = largest_millennium_falcon_set['theme_id'].values[0]

Q16_check = minifigs_from_themes(millennium_falcon_theme_id)

print(Q16_check)
print(f"Total Minifigs: {Q16_check.shape[0]}")

        fig_num                                           fig_name
0    fig-000003                              Assassin Droid, White
1    fig-000124                           Chewbacca, Reddish Brown
2    fig-000144  Luke Skywalker, White Robe, White Legs, Cheek ...
3    fig-000303           Clone Trooper, Phase I Armor, Brown Eyes
4    fig-000516  Darth Vader, Light Bluish Gray Skin, Scarred R...
..          ...                                                ...
144  fig-011942  AT-AT Driver - Light Bluish Grey Helmet, Sand ...
145  fig-011944  Snowtrooper, White Hips, Reddish Brown Head, F...
146  fig-011945  Snowtrooper, White Hips, Light Flesh Head, Female
147  fig-011946  Snowtrooper, White Hips, Light Flesh Head, Off...
148  fig-011957    Snowtrooper, White Hips, Light Flesh Head, Grin

[149 rows x 2 columns]
Total Minifigs: 149


**Q17:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [98]:
def sets_from_minifig(fig_num):

    inventory_ids = inventory_minifigs[inventory_minifigs['fig_num'] == fig_num]['inventory_id']
    
    sets_including_minifig = inventories[inventories['id'].isin(inventory_ids)]
    
    merged_df = pd.merge(sets_including_minifig, sets, left_on='set_num', right_on='set_num')
    
    result_df = merged_df[['set_num', 'version', 'name']].copy()
    result_df['fig_num'] = fig_num
    fig_name = minifigs[minifigs['fig_num'] == fig_num]['name'].values[0]
    result_df['fig_name'] = fig_name
    
    result_df.columns = ['set_num', 'version', 'name_sets', 'fig_num', 'fig_name']
    
    result_df = result_df.sort_values(by=['set_num', 'version']).reset_index(drop=True)
    
    return result_df

Q17_check = sets_from_minifig("fig-001039")

print(Q17_check)

   set_num  version            name_sets     fig_num  \
0  30100-1        1  Andrea on the Beach  fig-001039   
1  30107-1        1       Birthday Party  fig-001039   
2   3061-1        1       City Park Café  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...  


**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [100]:
Q17_check = sets_from_minifig("fig-001039")

print(Q17_check)

   set_num  version            name_sets     fig_num  \
0  30100-1        1  Andrea on the Beach  fig-001039   
1  30107-1        1       Birthday Party  fig-001039   
2   3061-1        1       City Park Café  fig-001039   

                                            fig_name  
0  Andrea - Bright Light Orange Top, Light Aqua S...  
1  Andrea - Bright Light Orange Top, Light Aqua S...  
2  Andrea - Bright Light Orange Top, Light Aqua S...  


**Q18:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

*You should only need to merge the following datasets for this answer so as to not go above the RAM limit: `themes`, `sets`, `inventories`, `inventory_parts`, `parts`, and `colors`*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [134]:
def set_parts(set_num, version_num):
    inventory_id = inventories[(inventories['set_num'] == set_num) & (inventories['version'] == version_num)]['id'].values[0]
    
    parts_in_inventory = inventory_parts[inventory_parts['inventory_id'] == inventory_id]
    
    parts_details = pd.merge(parts_in_inventory, parts, left_on='part_num', right_on='part_num', suffixes=('', '_part'))
    
    parts_details = pd.merge(parts_details, colors, left_on='color_id', right_on='id', suffixes=('', '_color'))
    
    parts_details = pd.merge(parts_details, inventories[['id', 'set_num', 'version']], left_on='inventory_id', right_on='id')
    
    parts_details = pd.merge(parts_details, sets[['set_num', 'name', 'year', 'theme_id']], left_on='set_num', right_on='set_num', suffixes=('', '_set'))

    parts_details = pd.merge(parts_details, themes[['id', 'name']], left_on='theme_id', right_on='id', suffixes=('', '_theme'))

    result_df = parts_details[['set_num', 'version', 'name_set', 'year', 'name_theme', 'part_num', 'name', 'name_color', 'quantity', 'is_spare']]
    result_df.columns = ['set_num', 'version', 'name_sets', 'year', 'name_themes', 'part_num', 'name_part', 'name_color', 'quantity', 'is_spare']
    
    result_df = result_df.sort_values(by=['part_num', 'name_color']).reset_index(drop=True)
    
    return result_df

millennium_falcon_set_num = largest_millennium_falcon_set['set_num'].values[0]

Q18_check = set_parts(millennium_falcon_set_num, 1)

print(f"Total rows: {Q18_check.shape[0]}")
print(Q18_check)

Total rows: 730
     set_num  version          name_sets  year                name_themes  \
0    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
1    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
2    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
3    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
4    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
..       ...      ...                ...   ...                        ...   
725  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
726  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
727  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
728  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
729  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   

    part_num                                          name_

**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in the `Millennium Falcon Prep Exercise`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [132]:
millennium_falcon_set_num = largest_millennium_falcon_set['set_num'].values[0]

Q18_check = set_parts(millennium_falcon_set_num, 1)

print(f"Total rows: {Q18_check.shape[0]}")
print(Q18_check)

Total rows: 730
     set_num  version          name_sets  year                name_themes  \
0    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
1    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
2    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
3    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
4    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
..       ...      ...                ...   ...                        ...   
725  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
726  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
727  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
728  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
729  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   

    part_num                                          name_

## Optional Exercises
These exercises are optional and are not graded by CodeGrade.  They can be used as extra optional practice.

- What is the median year for the sets data?  Given the oldest and newest year in the data, what does this median value tell you about whether the number of sets have increased or decreased over the years?
- What minifig has the most number of parts?
- What minifig shows up in the most number of sets?
- When was the first minifig included in a set?
- What set has the most number of minifigs
- What part is used the most for all sets? 
- What color is used the most for all parts?
- Get a list of all unique color names, sort by color name, and output as a list. 
- How many total transparent colors are in the colors DataFrame?
- Create a function called get_part_colors that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.
- What part / color combination is used the most for all sets?
- What part / color combination is used the least for all sets?
- Explore some of your favorite themes and which colors are used the most for those themes.
- Plot a line graph of the average number of parts per set per year.
